core/ordering/cosipbft/controller/action.go

```diff
@@ -166,31 +166,35 @@ func (rosterAddAction) Execute(ctx node.Context) error {
 		return xerrors.Errorf("injector: %v", err)
 	}
 
+	wait := ctx.Flags.Duration("wait")
+
+	// Start listening for new transactions before sending the new one, to
+	// be sure the event will be received.
+	watchCtx, cancel := context.WithTimeout(context.Background(), wait)
+	defer cancel()
+
+	events := srvc.Watch(watchCtx)
+
 	err = p.Add(tx)
 	if err != nil {
 		return xerrors.Errorf("failed to add transaction: %v", err)
 	}
 
-	wait := ctx.Flags.Duration("wait")
 	if wait > 0 {
-		err := waitTx(srvc, wait, tx)
-		if err != nil {
-			return xerrors.Errorf("wait: %v", err)
-		}
-	}
-
-	return nil
-}
-
-func waitTx(srvc Service, wait time.Duration, tx txn.Transaction) error {
```

Comment:
I'm not sure that joining these two functions again makes for a smoother read.
I'd split them the other way around though: one doing all the initialisation, which Execute() would call, and then doing the wait part inline. 
Either ways, this is okay. We can always refactor later if need be.